graduate SuspendJob to beta

adtac opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind api-change

#### What this PR does / why we need it:

1. Graduates SuspendJob to beta.
2. Adds a label to two existing Job metrics.
3. Changes the behaviour of syncJob so that each operation only either creates pods or deletes pods (but never both at the same time). This is okay because this will happen only in indexed Jobs with failed indexes, where we'll want to do both. And in such situations, we don't necessarily need to do both pod creations and deletions in the same sync.

KEP update: https://github.com/kubernetes/enhancements/pull/2646

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref: https://github.com/kubernetes/enhancements/issues/2232

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Suspend Job feature graduated to beta
Added the "action" label to Job controller sync metrics job_sync_total and job_sync_duration_seconds
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig apps
/cc @alculquicondor @ahg-g

Commit:
graduate SuspendJob to beta

Also adds a label to two existing Job metrics.

Signed-off-by: Adhityaa Chandrasekar <adtac@google.com>

alculquicondor commented on pkg/apis/batch/types.go:
```diff
@@ -199,9 +199,9 @@ type JobSpec struct {
-	// requires the SuspendJob feature gate to be enabled; otherwise this field
-	// may not be set to true. Defaults to false.
+	// resetting the ActiveDeadlineSeconds timer too. This field is beta-level.
+	// Requires the SuspendJob feature gate (enabled by default). Defaults to
```
Requires the SuspendJob feature gate to be enabled (enabled by default).

The "Defaults to false" sentence should be before the beta sentence. Also better put the beta sentence in a separate paragraph.

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -907,12 +906,10 @@ func (jm *Controller) manageJob(job *batch.Job, activePods []*v1.Pod, succeeded
 	}
 
-	return active, nil
+	return active, "tracking", nil
```
can you put the label values as constants in the metrics package we a comment for what they mean?

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -812,9 +813,7 @@ func (jm *Controller) manageJob(job *batch.Job, activePods []*v1.Pod, succeeded
-		if err != nil {
-			return active, err
-		}
+		return active, "pods_deleted", err
```
add to the description that we are changing the sync to only do deletions or creations but not both and why that is acceptable.

alculquicondor commented on pkg/controller/job/metrics/metrics.go:
```diff
@@ -37,21 +40,25 @@ var (
 			Buckets:        metrics.ExponentialBuckets(0.001, 2, 15),
 		},
-		[]string{"completion_mode", "result"},
+		[]string{"completion_mode", "result", "action"},
```
add this change to the release note

logicalhan replied:
Definitely, it's a breaking metrics change.

alculquicondor replied:
Not really breaking considering that this metric didn't hit a release yet.

alculquicondor commented on test/e2e/apps/job.go:
```diff
@@ -111,10 +108,7 @@ var _ = SIGDescribe("Job", func() {
 
-	// Requires the alpha level feature gate SuspendJob. This e2e test will not
-	// pass without the following flag being passed to kubetest:
-	//   --test_args="--feature-gates=SuspendJob=true"
```
Did you add it to the infra repo? Don't forget to remove it.

adtac replied:
yep, once this PR is merged

adtac replied:
ref: https://github.com/kubernetes/test-infra/pull/22188

ahg-g commented on pkg/apis/batch/types.go:
```diff
@@ -199,9 +199,11 @@ type JobSpec struct {
-	// may not be set to true. Defaults to false.
+	// resetting the ActiveDeadlineSeconds timer too. Defaults to false.
+	//
+	// This field is beta-level. Requires the SuspendJob feature gate to be
+	// enabled (enabled by default).
```
```suggestion
	// This field is beta-level, gated by SuspendJob feature flag.
```

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -812,9 +813,7 @@ func (jm *Controller) manageJob(job *batch.Job, activePods []*v1.Pod, succeeded
-		if err != nil {
-			return active, err
-		}
+		return active, metrics.JobSyncActionPodsDeleted, err
```
add a comment indicating that we either delete or add pods in a cycle, and deleting pods takes precedence.

ahg-g replied:
this wasn't addressed

adtac replied:
ah, must have missed it. done.

ahg-g commented on pkg/controller/job/job_controller.go:
```diff
@@ -824,7 +823,7 @@ func (jm *Controller) manageJob(job *batch.Job, activePods []*v1.Pod, succeeded
 		}
 		if diff == 0 {
-			return active, nil
+			return active, metrics.JobSyncActionTracking, nil
```
how do the above two conditions ever be possible if wantActive is strictly larger than active in this branch?

alculquicondor replied:
they are not possible, but the code was there 🤷‍♂️

alculquicondor replied:
that said, let's remove it.

alculquicondor commented on pkg/controller/job/metrics/metrics.go:
```diff
@@ -63,6 +70,25 @@ var (
 	)
 )
 
+var (
```
why not const?

logicalhan commented:
/triage accepted
/assign

alculquicondor commented:
maybe still needs rebase?

adtac commented:
Not sure what's changed, but I had to run `update-generated-swagger-docs.sh` and `update-openapi-spec.sh` instead of `make generated_files` for the changes to be picked up for some reason o.O

Anyway, rebased and squashed.

alculquicondor commented:
/lgtm
/label api-review

alculquicondor commented:
/assign @soltysh

adtac commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

ahg-g commented:
/lgtm

Just one final nit.

adtac commented:
/retest

ahg-g commented:
/assign @janetkuo

soltysh commented on pkg/controller/job/job_controller.go:
```diff
@@ -812,21 +813,15 @@ func (jm *Controller) manageJob(job *batch.Job, activePods []*v1.Pod, succeeded
 
 	if active < wantActive {
 		diff := wantActive - active
-		if diff < 0 {
```
Any particular reason to drop this safety net?

alculquicondor replied:
but it's already checked in line 820

ahg-g replied:
because of the condition above, `wantActive` is strictly larger than `active`

ahg-g replied:
ha, I should have refreshed my page, I didn't see your replies :)

soltysh commented on pkg/controller/job/metrics/metrics.go:
```diff
@@ -27,8 +27,11 @@ import (
+	// values:
+	//   completion_mode: Indexed, NonIndexed
+	//   result:          success, error
+	//   action:          reconciling, tracking, pods_created, pods_deleted
```
I haven't seen approval from the monitoring team on this change, yet.

alculquicondor replied:
@logicalhan PTAL

ahg-g replied:
@logicalhan can you please review the metrics changes?

janetkuo reviewed (approved):
LGTM from my end; need @logicalhan to look at the metrics change still.

logicalhan reviewed (commented):
/lgtm
From a metrics perspective. When this feature is goes to GA, you may want to consider promoting this metric to STABLE, since presumably people will be alerting off of it.

liggitt commented:
/approve

for API doc update

liggitt commented:
/hold in case any other reviews were desired (feel free to self-unhold if not)

alculquicondor commented:
all parties already LGTMed

/hold cancel
